### ISSUE TITLE
Firebase issues

### ISSUE THREAD

**Issue by dcristolovean** [Subito-it/PodBuilder#8]
I have a rather complex project with a lot of pods and I'm trying to fix the issues one by one as I run into them. Now, Firebase time :)

To (kinda) get firebase to run (and we're using a lot of stuff from them), I left "Firebase" in force_prebuild_pods.  All good so far, the frameworks built but I had to change a lot of import Firebase to import FirebaseXXX modules. 
Managed to do that and get the project compiling, at least until the Run Script build phase.

FirebaseCrashlytics needs to run a "run" script from the old pods folder, which is not possible now if it's a framework.

Any ideas what I can do ?

- I tried putting Firebase/Crashlytics in the skip_pods: no luck, it's still being built
- I tried putting ALL firebase stuff I have in the skip_pods: no luck, all are built again.

**Comment by tcamin** [comment]
> I tried putting ALL firebase stuff I have in the skip_pods: no luck, all are built again.
try again with beta 6. I tried on a sandbox project and it works as expected, watch out the the same pods are not specified in the `force_prebuild_pods` configuration.

> the frameworks built but I had to change a lot of import Firebase to import FirebaseXXX modules.
Import with the origianl module name should work without renaming if you use the beta version of PodBuilder.

> FirebaseCrashlytics needs to run a "run" script from the old pods folder, which is not possible now if it's a framework.
This is indeed an issue. In our project we didn't include the run script since we manually upload dSYMs (including the ones of the prebuilt frameworks) manually when submitting to the AppStore.
Alternatively you can manually extract the executable (e.g. via wget from Firebase's GitHub repo), commit it in the repo and create a manual run script. Double check if you really need it, because it is important to manually upload the dSYMs of the prebuilt items to properly symbolicate crash stacktraces anyway.

**Comment by dcristolovean** [comment]
Can you please take a look at Firebase ? The behavior is very weird:  (all tests are done with clean Pods/ and Frameworks/ folders)

1) Put them in skip_pods (and only there) -> build_all builds all of them in Frameworks and also downloads them in Pods
2) Take them out from skip_pods, put Firebase in force_prebuild_pods -> build_all builds them all again, but also the root Podfile is not updated to use them, they are also still downloaded in Pods/ .

Basically, I got to a state where they are always built, never used in the root Podfile (doesn't point to them, the Firebase area is unchanged). Any ideas?

**Comment by dcristolovean** [comment]
One weird one... running a sync pods after I added a new target in Frameworkd/Podfile... and those Firebase entries you see above, they all changes to PromisesObjc (or ProtoBuf, don't remember exactly). 
So 7 identical entries, all pointing to one dependency of Firebase :)

Something's up with the code when it deals with Firebase, guess it still needs some work.

**Comment by tcamin** [comment]
The issue with skip_pods should be fixed in beta 8. I could bot replicate the problem with the PromisesObjc/ProtoBuf. Can you try again with latest version to see if it is still broken?

**Comment by dcristolovean** [comment]
skip_pods seems ok and the other problem didn't reproduce either.
But also, sync pods doesn't really do what I want...  If I add a new target in Frameworks/Podfile, I'd like to assign all current frameworks/Pods to it, sort of a pod install.  (without any update and so on).
Sync pods seems to update the root Podfile indeed, but that's about it.

**Comment by dcristolovean** [comment]
And Beta 8 introduced a bit problem, I tested 3-4 times:

So I had my master branch done with Beta 7, all good, working.  I installed Beta 8, run a build_all and it suddenly introduced a dependency to a framework called AWSRuntime 
- I didn't change anything on my side, nor added new deps
- I don't think this AWSRuntime even exists and there isn't any framework built for it.
- code compilation fails now because of this.

I'm kinda stuck now, so, how can I put back Beta 7, even without the skip_pods working ?

**Comment by dcristolovean** [comment]
Installing AWSS3 2.14.1 (was 0.0.2) ...

after a few more tries :)

**Comment by tcamin** [comment]
As far as I understood the AWS problems were related to a bad update of some podspecs. So to recap, what are the current open problems?

**Comment by dcristolovean** [comment]
- AWSRuntime is not correctly built now (not built at all actually, but added in the Other Flags in the project, in all targets) -> compile failed.    It's fixed if you manually put pod 'AWSRuntime' in the Podfile
- even if the PromisesObjc is not built (it's a dependency of Firebase), the podspec file is always created in Frameworks (just for this one, not for any other dependency)
- Not sure why AWSS3 was installed as 0.0.2 (this is actually the version of AWSRuntime), but after putting AWSRuntime the 1st pod in the Podfile before any other AWS pods, it fixed itself, as you can see, to 2.14.1.   Weird.
- sync_podfile ... I need to look more into this, but it still doesn't seem to do what I want.  
We're adding new targets via some Fastlane lanes and the final step was to run a pod install.  That would actually ensure the last target added was linked to all the pods too and none of the pods was actually updated. 
Now, I replaced that pod install with pod_builder sync_podfile, but as I saw, this doesn't actually link my new target so it doesn't compile. 
I need to do a build_all, which also updates everything, which is not something I might want in the future.
Basically...how do I make sure my new target works, with the current built frameworks and already downloaded Pods (not prebuilt ones) ?

Guess that covers everything for now, until next time :)

**Comment by dcristolovean** [comment]
And another very very weird one:

I use a Fastlane lane to duplicate a target and create a new one. At the end, there's a sh called that (used to do a pod install at the end, to update the new target with all the Pods) is doing now a pod_builder build_all.  (since sync_podfile doesn't do anything).  I am in the root folder, but the build_all command fails from here, with the old error about SWIFT_VERSION in 3 of my Pods.

If I run manually a pod_builder build_all.... it works fine.

Any ideas what might confuse it if run from a script ?

**Comment by tcamin** [comment]
> even if the PromisesObjc is not built (it's a dependency of Firebase), the podspec file is always created in Frameworks (just for this one, not for any other dependency)

The fact that there are podpsec on non prebuilt frameworks shouldn't be seen as a problem. Those podspecs are autogenerated and should be treated as internal implementation of PodBuilder.

> Not sure why AWSS3 was installed as 0.0.2 (this is actually the version of AWSRuntime), but after putting AWSRuntime the 1st pod in the Podfile before any other AWS pods, it fixed itself, as you can see, to 2.14.1. Weird.

Can you list the AWS pod you're using in the "wrong order" so that I can replicate the issue?

> sync_podfile ... I need to look more into this, but it still doesn't seem to do what I want.

sync_podfile takes the Podfile in the Framework folder and updates the Application's Podfile accordingly as per documentation. When you add a new target the associated pods won't be automatically converted to a prebuilt entries. They just match the original Podfile

If you know some of those pods are already prebuilt in other targets you can use the `switch` command (after the sync_podfile) to use the prebuilt version.

There's also the info command that gives you the detailed list of prebuilt elements which could be useful to find out which pods are already prebuilt.

**Comment by dcristolovean** [comment]
- yeah, no big deal, but I deleted what I don't need and it bugs me when I have to git add that file :) Others are not created, just this one. Eh, no biggie.
- they are listed in a post above, that's the order.  I put AWSRuntime first now and it's ok
- yes, I seems that's what it does.  
But you know, if you have some (normal, not prebuilt) pods installed and you just add a new target, you can run 'pod install' and it won't do anything except add all the pods to the new target and create the pod project for it ?
I'm looking to do now the same thing with pod_builder
This will ensure my new target works and also should NOT update or get any new versions for any prebuilt or not Pod.
I want to control that and to real testing when I update pods, can't just do it everytime we have a new target.  This is a sort of "template" app that's customized for each client and so on, so adding new targets is quite common.

**Comment by dcristolovean** [comment]
An updated list of issues now:

- the question above is still pending:  how do I do the equivalent of the 'old' pod install when I add a new target, without updating any Pod or pre-built framework, so just linking stuff to the new target and creating the pods project

- Firebase.. man, this kills me.  I have a very stable setup now, all prebuilt frameworks except Firebase and a lot of Google shit.
I tried again to give it a go to Firebase.
So:   I took the firebase pods from the "skip_pods" section... put "Firebase" in the "force_prebuild_pods"... run a build_all.
The result...  "some" Firebase stuff was pre-built (not all), they can be seen in Frameworks/  but all the Pods are still there where they were, they were not removed, the Podfile was not updated (as if firebase was never prebuilt).  The app compiles but instantly crashes at startup with errors too wierd and many to even mention.
The entire Firebase prebuilding process is not working well at all.   This are my pods:

    pod 'Firebase'
    pod 'Firebase/Core'
    pod 'Firebase/Database'
    pod 'Firebase/Auth'
    pod 'Firebase/Firestore'
    pod 'Firebase/Performance'
    pod 'Firebase/Crashlytics'

If I could get these 2 things to work, the new target and Firebase, everything would be perfect.

**Comment by tcamin** [comment]
> the question above is still pending: how do I do the equivalent of the 'old' pod install when I add a new target, without updating any Pod or pre-built framework, so just linking stuff to the new target and creating the pods project

Update the PodBuilder-Podfile (under Frameworks) adding target, run `pod_builder sync_podfile`, then add some logic on your side that decides which of the pods in the new target should be used as prebuild. For each of them just call `pod_builder switch --prebuilt <pod_name>`. 

> Firebase

The Firebase spec is a bit twisted instead of including the subspec did you try to include the spec the subspec points to?

pod 'Firebase'
pod 'FirebaseCore'
pod 'FirebaseDatabase'
pod 'FirebaseAuth'
pod 'FirebaseFirestore'
pod 'FirebasePerformance'
pod 'FirebaseCrashlytics'

**Comment by dcristolovean** [comment]
This leads to weird things:

Downloading dependencies
Installing AFNetworking (4.0.1)
Installing AWSAPIGateway (2.14.2)
Installing AWSAuthCore (2.14.2)
Installing AWSAuthUI (2.14.2)
Installing AWSCognito (2.14.2)
Installing AWSCognitoIdentityProvider (2.14.2)
Installing AWSCognitoIdentityProviderASF (1.0.1)
Installing AWSCore (2.14.2)
Installing AWSMobileClient (2.14.2)
Installing AWSRuntime (0.0.2)
Installing AWSS3 (2.14.2)
Installing AWSSNS (2.14.2)
Installing AWSUserPoolsSignIn (2.14.2)
Installing BoringSSL-GRPC (0.0.7)
Installing DLRadioButton (1.4.12)
Installing DailymotionPlayerSDK (3.7.8)
Installing DateTools (2.0.0)
Installing FBSDKCoreKit (7.1.1)
Installing FBSDKLoginKit (7.1.1)
Installing FBSDKShareKit (7.1.1)
Installing Firebase (6.28.1)
Installing FirebaseAnalytics (6.6.2)
Installing FirebaseAuth (6.7.0)
Installing FirebaseCore (6.9.1)
Installing FirebaseCoreDiagnostics (1.5.0)
Installing FirebaseCrashlytics (4.3.0)
Installing FirebaseDatabase (6.4.0)
Installing FirebaseFirestore (1.16.2)
Installing FirebaseInstallations (1.5.0)
Installing GTMSessionFetcher (1.4.0)
Installing GoogleAppMeasurement (6.6.2)
Installing GoogleDataTransport (7.0.0)
Installing GoogleUtilities (6.7.1)
Installing Harpy (4.1.14)
Installing ImageSizeFetcher (0.9.1)
Installing JSQMessagesViewController (7.2.0)
Installing JSQSystemSoundPlayer (2.0.1)
Installing JVFloatLabeledTextField (1.2.2)
Installing KYDrawerController (2.1.0)
Installing Kingfisher (5.14.1)
Installing Koloda (5.0.1)
Installing LocalizedTimeAgo (1.3.0)
Installing MBProgressHUD (1.2.0)
Installing PromisesObjC (1.2.9)
Installing Pulley (2.8.0)
Installing SDWebImage (4.4.8)
Installing XCDYouTubeKit (2.14.0)
Installing abseil (0.20200225.0)
Installing gRPC-C++ (1.28.2)
Installing gRPC-Core (1.28.2)
Installing leveldb-library (1.22)
Installing nanopb (1.30905.0)
Installing pop (1.0.12)
Generating Pods project
Skipping User Project Integration
Building frameworks
Built 100 frameworks
Copying 104 frameworks to `Rome`
Pod installation complete! There are 108 dependencies from the Podfile and 53 total pods installed.
Cleaning framework folder
Writing licenses
Generating PodBuilder's local podspec
Writing Restore Podfile
Analyzing dependencies
Pre-downloading: `JSQMessagesViewController` from `https://github.com/dcristolovean/JSQMessagesViewController.git`, commit `d62b6388440766f636bdb075e4c8ff9e4627dea6`
Updating Application Podfile
Running pod install
Analyzing dependencies
[!] CocoaPods could not find compatible versions for pod "abseil/strings/strings":
  In Podfile:
    FirebaseFirestore (from `Frameworks/Rome`) was resolved to 1.16.2, which depends on
      abseil/strings/strings

None of your spec sources contain a spec satisfying the dependency: `abseil/strings/strings`.

You have either:
 * out-of-date source repos which you can update with `pod repo update` or with `pod install --repo-update`.
 * mistyped the name or version.
 * not added the source repo that hosts the Podspec to your Podfile.


🎉 done!

**Comment by tcamin** [comment]
Does running pod repo update help?

**Comment by dcristolovean** [comment]
no, still getting 

[!] CocoaPods could not find compatible versions for pod "abseil/strings/strings":
  In Podfile:
    FirebaseFirestore (from `Frameworks/Rome`) was resolved to 1.16.2, which depends on
      abseil/strings/strings

None of your spec sources contain a spec satisfying the dependency: `abseil/strings/strings`.

**Comment by tcamin** [comment]
Does a normal Podfile with CocoaPods work?

**Comment by dcristolovean** [comment]
yes, all good with the old 'normal' Podfile.

**Comment by tcamin** [comment]
I have good news wrt to this issue. I have made many changes lately to the 2.x codebase and it is finally properly handing the abseil podspec. This means the Firebase setup mentioned before in now properly working, see the Example Podfile (https://github.com/Subito-it/PodBuilder/blob/deterministic_build/Example/Frameworks/Podfile).

**Comment by dcristolovean** [comment]
Unfortunately, no, I don't see any improvements in my case.  Here's what happens now:

- installed the latest beta
- in the json file, took out Firebase from the "skip_pods"
- put Firebase in force_prebuild_pods, not sure if this is even necessary
- run a pod_builder build_all

Issues:

- you added a sha1sum dependency, which is not a standard tool on MacOS. shasum is.  Anyway, easy fix, I installed it, no more errors (maybe update the README?)
- Firebase stuff is being built and its dependecies. They appear in /Frameworks, correctly.
- The Podfile (root folder) does NOT get updated to use these new Frameworks
- Firebase Pods are still left in /Pods
- Project is NOT updated to use the newly build stuff in /Frameworks for Firebase, but something is modified inside
- Trying to run the app, XCode fails, cannot find a class named "#$#@$".  Well, not exactly those chars, but something close to that.  Yes, I'm not kidding, XCode tries to compile a class with a very weird name and of course fails. That class cannot be found anywhere.

That's about it, the only thing that works is the build of the Firebase frameworks, but removing the old ones and correctly integrating the new ones in the proj doesn't work at all.

**Comment by tcamin** [comment]
> you added a sha1sum dependency, which is not a standard tool on MacOS. shasum is. Anyway, easy fix, I installed it, no more errors (maybe update the README?)

This will be fixed in the next release.

> Firebase stuff is being built and its dependecies. They appear in /Frameworks, correctly.

This wasn't working before right?

Did you try building the example project shipped with PodBuilder?

Using beta 19:
```
git clone https://github.com/Subito-it/PodBuilder.git
cd PodBuilder
git checkout deterministic_build
pod_builder build_all 
```

I see the Podfile of the app being correctly updated and the app builds correctly.

**Comment by dcristolovean** [comment]
This thing with Firebase being build but still left in Pods happened before.  At this point, I see nothing new with this last version, I think I already got all those issues at some point or another.

**Comment by tcamin** [comment]
Did you check if the Firebase folder in Pods contains source code or just headers/

**Comment by dcristolovean** [comment]
Unfortunately no.  Deleted that test branch too, so I can only check when I try again.
Either way, they should be removed completely from there, no ?

**Comment by dcristolovean** [comment]
Big issue again with Firebase... same old setup that worked, didn't touch it for weeks.. 
Now I added a new target, run pod_builder build_all (because I don't have any other way to just link the new target to the existing stuff.........!) and I'm getting this:

[!] CocoaPods could not find compatible versions for pod "FirebaseDatabase":
  In snapshot (Podfile.lock):
    FirebaseDatabase

  In Podfile:
    Firebase/Database was resolved to 6.22.0, which depends on
      FirebaseDatabase (~> 6.1.4)

    FirebaseDatabase (from `Frameworks/Prebuilt/FirebaseDatabase`)

Any ideas ? Quite urgent....

**Comment by tcamin** [comment]
Which version of pod_builder?
Do you have `pod 'Firebase/Database'` or `pod 'FirebaseDatabase'`?

**Comment by dcristolovean** [comment]
latest beta, just installed today.

Podfile:

pod 'Firebase'
    pod 'Firebase/Core'
    pod 'Firebase/Database'
    pod 'Firebase/Auth'
    pod 'Firebase/Firestore'
    pod 'Firebase/Performance'
    pod 'Firebase/Crashlytics'

Podbuilder.json:

"skip_pods": [
    "GoogleMaps",
    "Firebase",
    "Firebase/Core",
    "Firebase/Database",
    "Firebase/Auth",
    "Firebase/Firestore",
    "Firebase/Performance",
    "Firebase/Crashlytics"
  ],

**Comment by tcamin** [comment]
As stated previously did you try replacing subspec with their associated specs?

pod 'FirebaseCore'
pod 'FirebaseDatabase'
pod 'FirebaseAuth'
pod 'FirebaseFirestore'
pod 'FirebasePerformance'
pod 'FirebaseCrashlytics'

Podbuilder.json:

"skip_pods": [
"GoogleMaps",
"Firebase"
],

**Comment by dcristolovean** [comment]
worked like that it seemed.  Although 4 Firebase subprojects were removed now and one of them was not correctly removed from Other Linker Flags and I had to do it by hand.

**Comment by tcamin** [comment]
It would be helpful if you could provide more context to see if this is somehow fixable

**Comment by dcristolovean** [comment]
Dead in the water again. This time the culprit is FirebaseCore.  Latest beta for pod-builder.  Latest cocoapods 1.10 rc1.
Same project that worked last time, now I needed to add a new target.

1) if I put FirebaseCore in skip_pods, **It will install v 6.10.3**
-> all builds well, but getting a compile error after that:


Undefined symbols for architecture x86_64:
  "_GULResetLogger", referenced from:
      _FIRResetLogger in FirebaseCore(FIRLogger.o)
ld: symbol(s) not found for architecture x86_64



2) Take FirebaseCore out of skip pods, so it's prebuilt...  (which is what I always did and it worked)
**It will install v 6.10.4**

But throw this:

 CocoaPods could not find compatible versions for pod "FirebaseCore":
  In Podfile:
    Firebase was resolved to 6.33.0, which depends on
      Firebase/Core (= 6.33.0) was resolved to 6.33.0, which depends on
        Firebase/CoreOnly (= 6.33.0) was resolved to 6.33.0, which depends on
          FirebaseCore (= 6.10.3)

    Firebase was resolved to 6.33.0, which depends on
      Firebase/Core (= 6.33.0) was resolved to 6.33.0, which depends on
        FirebaseAnalytics (= 6.8.3) was resolved to 6.8.3, which depends on
          FirebaseCore (~> 6.10)

Take note of the bold stuff, it install different versions of FirebaseCore, based on it being included in skip pods or not.

Also, all there problems occur for me because there is no way to just add a target to the project now and not do a full build_all, which gets the latest pods uncontrolled.
I just need an option so pod builder can just link a new target, without updating any pod.
It's exactly what a simple 'pod install' is doing.

**Comment by tcamin** [comment]
PodBuilder always rebuilds latest version fetched from specs.

This is what happens when FirebaseCore is not added to skipPods. On the other hand when you add it to the skipPods it's added to the application Pods as a normal pod. This means that the cocoapods resolves version honoring Podfile.lock, and this is probably the reason because the 2 versions do not match. My advice, and this is what we use in our project, is to pin versions in the PodBuilder Podfile of all pods (eg. `pod 'FirebaseCore', '=6.10.3'`).

With regards to your problem you should use `pod_builder sync_podfile`. I just tried the following.
1. sample project with 1 target Target1
2. added a pod to Target1 (`pod 'AFNetworking'`).
3. build_all. The Application podfile points to prebuilt version of AFNetworking.
4. Add new target Target2
5. Add Target2 entry to the PodBuilder/Podfile with the same pod (`pod 'AFNetworking'`).
6. `pod_builder sync_podfile`
7. The Application podfile now contains the 2 targets both with the AFNetworking pod pointing to the prebuilt versions.

**Comment by dcristolovean** [comment]
I'm gonna try setting versions in my Podfile to avoid unwanted updates.

But, regarding the sync_podfile command, that really doesn't work as you say.
I am just trying now and it's just installing ALL pods in /Pods.   Even AFNetworking is being installed in /Pods, although I already had it in Prebuilt.
This is happening because the root Podfile is identical now to the original Pod, it lost all the pointers to the Prebuilt folders, it's just a normal Podfile so I guess it's just doing a normal full pod install.

So... instead of using the Prebuild/Podfile and do it's stuff, prebuilding or using the already prebuilt apps.... it just copied to the root folder and it's doing a normal pod install..

Help :)

**Comment by dcristolovean** [comment]
hold on to that, I redid the flow and run the sync_podfile by hand, instead from the script where it was and now it didn't get all my Pods again.
This is very weird.
